compiler-rt/lib/ctx_profile/CtxInstrProfiling.cpp

```diff
@@ -418,12 +436,24 @@ void __llvm_ctx_profile_start_collection() {
       resetContextNode(*Root->FirstUnhandledCalleeNode);
     __sanitizer::atomic_store_relaxed(&Root->TotalEntries, 0);
   }
+  if (AutodetectDuration) {
+    auto *RD = new (__sanitizer::InternalAlloc(sizeof(RootAutoDetector)))
+        RootAutoDetector(AllFunctionsData, RootDetector, AutodetectDuration);
+    RD->start();
+  } else {
+    __sanitizer::Printf("[ctxprof] Initial NumMemUnits: %zu \n", NumMemUnits);
+  }
   __sanitizer::atomic_store_relaxed(&ProfilingStarted, true);
```

Comment:
Is relaxed the right choice? Is there an implication of other threads observing ProfilingStarted as true before the call to RD->start() completes?

Is this why you have the RD->join() code below?

Reply:
RD->start just sets up the waiting thread. Threads observing `ProfilingStarted` and `RootDetector` just insert sample stacks in the heap. When eventually that waiting thread started, and it finished waiting, it collects whatever threads inserted in the interim.